### PR TITLE
Cancellation for `CreateNewImage` task

### PR DESCRIPTION
**Pull request by vlada-shubina**
fixes https://github.com/dotnet/sdk-container-builds/issues/231

- implemented `ICancelableTask` for  `CreateNewImage`
- passing cancellation tokens all the way down
- fixed naming of async methods (`Async` suffix)
- `ILocalDocker.IsAvailable` - sync implementation for use in tests

### CODE REVIEW

**Comment by baronfel** [comment]
> ILocalDocker.IsAvailable - sync implementation for use in tests

Ah, your nemesis at last is slain :D

File: containerize/Program.cs [hunk 1]
```diff
@@ -200,7 +200,24 @@
...
+        _name,
+        _tags,
+        _outputReg,
+        _labels,
+        _ports,
+        _envVars,
+        _rid,
+        _ridGraphPath,
+        _localContainerDaemon,
+        context.GetCancellationToken()).ConfigureAwait(false);
```
Comment (vlada-shubina) [inline containerize/Program.cs:220]: I think we need https://github.com/dotnet/sdk-container-builds/issues/172 to get cancellation working. The cancellation middleware might not be available by default.
I need to dive into `ToolTask` implementation to see how it is cancelled.

**Review by baronfel — APPROVED** [review]
This LGTM - were you able to verify that cancelling the build from the full-framework build flowed through to the CLI as expected?

File: Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs [hunk 2]
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.NET.Build.Containers.Tasks;
 /// <summary>
 /// This task will shell out to the net7.0-targeted application for VS scenarios.
 /// </summary>
-public partial class CreateNewImage : ToolTask
+public partial class CreateNewImage : ToolTask, ICancelableTask
```
Comment (rainersigwald) [inline Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: `ToolTask` already implements `ICancelableTask` so this shouldn't be necessary, but the default ToolTask behavior is to just rudely kill the process. Does that work ok for us?

Comment (vlada-shubina) [reply Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: just discovered the same
- i kept `ICancelableTask` for readability only, so noone asks why it's not cancelable.
- i just discovered same about `ToolTask` and that's not good. This PR adds only very basic cancelation, and nothing really depends on it being triggered now. However, built-in methods might have better behavior with disposing / removing resources before cancellation happens. Ideally, ToolTask should send terminate a process (at least attempt to), before killing the process.

Comment (vlada-shubina) [reply Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: i see that `ToolTask` allows to override `Cancel` and `Execute`, but it feels like too much duplicate logic just to handle termination differently. Ideally `TerminateToolProcess` is the place to customize, but it's private now.

Also, adding an option to terminate the process might be a nice improvement overall for any `ToolTask` implementation.

Comment (rainersigwald) [reply Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: > Also, adding an option to terminate the process might be a nice improvement overall for any `ToolTask` implementation.

What do you mean by "terminate"? I don't think Windows or .NET has any equivalent of `SIGTERM`.

Comment (vlada-shubina) [reply Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: In windows there is a way to send Ctrl+C using https://learn.microsoft.com/en-us/windows/console/generateconsolectrlevent. 

https://stackoverflow.com/questions/813086/can-i-send-a-ctrl-c-sigint-to-an-application-on-windows has some details.

File: containerize/Program.cs (hunk 1 above)
Comment (vlada-shubina) [reply containerize/Program.cs:220]: `ToolTask` kills the process, so cancellation unfortunately is not happening.

**Comment by vlada-shubina** [comment]
> This LGTM - were you able to verify that cancelling the build from the full-framework build flowed through to the CLI as expected?

No, the cancelling the build won't terminate the process gracefully, it kills it. See https://github.com/dotnet/sdk-container-builds/pull/348#discussion_r1108772198 for more details.

File: Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs (hunk 2 above)
Comment (vlada-shubina) [reply Microsoft.NET.Build.Containers/Tasks/CreateNewImageToolTask.cs:16]: Note: System.CommandLine also adds handler for https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.processexit?view=net-7.0, but it is not called when process is killed either.

https://github.com/dotnet/command-line-api/blob/1bc03fed713b0b10455942d9f8eeab371d519cba/src/System.CommandLine/Builder/CommandLineBuilderExtensions.cs#L34-L121

**Pull request merged**